File: win/DoorbellApp/Core/MjpegStreamer.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using System.Threading;
using System.Windows.Media.Imaging;

namespace DoorbellApp.Core
{
    // Bounds on frame and header sizes limit untrusted multipart input. Frozen BitmapImage values
    // may be handed safely from the reader thread to the WPF dispatcher.
    public sealed class MjpegStreamer
    {
        private const int MaxFrame = 4 * 1024 * 1024;

        private readonly string _url;
        private readonly Action<BitmapImage, int> _onFrame;
        private readonly bool _lowResource;
        private volatile bool _running;
        private Thread _thread;

        public MjpegStreamer(string url, Action<BitmapImage, int> onFrame,
                             bool lowResource = false)
        {
            _url = url;
            _onFrame = onFrame;
            _lowResource = lowResource;
        }

        public void Start()
        {
            if (_running) return;
            _running = true;
            _thread = new Thread(Loop) { IsBackground = true, Name = "mjpeg" };
            _thread.Start();
        }

        public void Stop()
        {
            _running = false;
            try { _thread?.Interrupt(); } catch { }
            _thread = null;
        }

        private void Loop()
        {
            while (_running)
            {
                try
                {
                    var req = (HttpWebRequest)WebRequest.Create(_url);
                    req.Timeout = 4000;
                    req.ReadWriteTimeout = 10000;
                    using (var resp = req.GetResponse())
                    using (var raw = resp.GetResponseStream())
                    using (var ins = new BufferedStream(raw, 64 * 1024))
                    {
                        DateTime nextDecode = DateTime.MinValue;
                        while (_running)
                        {
                            int rotation;
                            byte[] frame = ReadPart(ins, out rotation);
                            if (frame == null) break;
                            if (_lowResource && DateTime.UtcNow < nextDecode) continue;
                            if (_lowResource)
                                nextDecode = DateTime.UtcNow.AddMilliseconds(250);
                            var bmp = Decode(frame, _lowResource ? 640 : 0);
                            if (bmp != null && _running) _onFrame(bmp, rotation);
                        }
                    }
                }
                catch {  }
                if (!_running) break;
                try { Thread.Sleep(2000); } catch (ThreadInterruptedException) { }
            }
        }

        public static BitmapImage Decode(byte[] jpeg)
        {
            return Decode(jpeg, 0);
        }

        public static BitmapImage Decode(byte[] jpeg, int maximumPixelWidth)
        {
            try
            {
                var bmp = new BitmapImage();
                using (var ms = new MemoryStream(jpeg))
                {
                    bmp.BeginInit();
                    bmp.CacheOption = BitmapCacheOption.OnLoad;
                    if (maximumPixelWidth > 0) bmp.DecodePixelWidth = maximumPixelWidth;
                    bmp.StreamSource = ms;
                    bmp.EndInit();
                }
                bmp.Freeze();
                return bmp;
            }
            catch { return null; }
        }

        private static byte[] ReadPart(Stream ins, out int rotation)
        {
            int contentLength = -1;
            rotation = 0;
            for (;;)
            {
                string line = ReadLine(ins);
                if (line == null) return null;
                if (line.Length == 0)
                {
                    if (contentLength > 0) break;
                    continue;
                }
                int c = line.IndexOf(':');
                if (c > 0 && line.Substring(0, c).Trim()
                        .Equals("Content-Length", StringComparison.OrdinalIgnoreCase))
                {
                    int v;
                    if (!int.TryParse(line.Substring(c + 1).Trim(), out v) ||
                        v <= 0 || v > MaxFrame)
                        return null;
                    contentLength = v;
                }
                else if (c > 0 && line.Substring(0, c).Trim()
                         .Equals("X-Doorbell-Video-Rotation", StringComparison.OrdinalIgnoreCase))
                {
                    int.TryParse(line.Substring(c + 1).Trim(), out rotation);
                    rotation = ((rotation % 360) + 360) % 360;
                }
            }
            var buf = new byte[contentLength];
            int off = 0;
            while (off < contentLength)
            {
                int n = ins.Read(buf, off, contentLength - off);
                if (n <= 0) return null;
                off += n;
            }
            return buf;
        }

        private static string ReadLine(Stream ins)
        {
            var sb = new StringBuilder(64);
            for (;;)
            {
                int ch = ins.ReadByte();
                if (ch < 0) return null;
                if (ch == '\n') return sb.ToString().TrimEnd('\r');
                sb.Append((char)ch);
                if (sb.Length > 512) return null;
            }
        }
    }
}

File: android/app/src/main/java/jp/ox/doorbell/MjpegStreamer.kt
// Bounded multipart MJPEG reader. Core's server supplies Content-Length on each part, allowing
// exact frame extraction without a third-party parser. Callbacks run on the reader thread.
package jp.ox.doorbell

import android.graphics.Bitmap
import android.os.SystemClock
import android.util.Log
import java.io.BufferedInputStream
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL

class MjpegStreamer(private val url: String, private val onFrame: (Bitmap, Int) -> Unit) {

    private data class Frame(val jpeg: ByteArray, val rotation: Int)

    @Volatile
    private var running = false
    private var thread: Thread? = null
    private var startupMs = 0L
    private var tracedHeaders = false
    private var tracedPart = false
    private var tracedDecode = false

    fun start() {
        if (running) return
        running = true
        startupMs = SystemClock.elapsedRealtime()
        tracedHeaders = false
        tracedPart = false
        tracedDecode = false
        trace("request_start")
        thread = Thread({ loop() }, "mjpeg-$url").apply {
            isDaemon = true
            start()
        }
    }

    fun stop() {
        running = false
        thread?.interrupt()
        thread = null
    }

    private fun loop() {
        while (running) {
            var conn: HttpURLConnection? = null
            try {
                conn = URL(url).openConnection() as HttpURLConnection
                conn.connectTimeout = 4000
                conn.readTimeout = 10000
                val ins = BufferedInputStream(conn.inputStream, 64 * 1024)
                if (!tracedHeaders) {
                    tracedHeaders = true
                    trace("http_response_headers")
                }
                while (running) {
                    val frame = readPart(ins) ?: break
                    if (!tracedPart) {
                        tracedPart = true
                        trace("first_complete_multipart")
                    }
                    val bmp = BoundedBitmapDecoder.decode(frame.jpeg, MAX_WIDTH, MAX_HEIGHT) ?: continue
                    if (!tracedDecode) {
                        tracedDecode = true
                        trace("first_jpeg_decoded")
                    }
                    if (running) onFrame(bmp, frame.rotation)
                }
            } catch (e: Exception) {
                if (running) Log.w(TAG, "stream error: $e")
            } finally {
                try { conn?.disconnect() } catch (_: Exception) { }
            }
            if (!running) break
            try { Thread.sleep(2000) } catch (_: InterruptedException) { }
        }
    }

    private fun trace(stage: String) {
        Log.i(TAG, "startup $stage +${SystemClock.elapsedRealtime() - startupMs}ms")
    }

    /** Read one bounded part; return null at the stream boundary. */
    private fun readPart(ins: InputStream): Frame? {
        var contentLength = -1
        var rotation = 0
        // The blank line after the multipart headers begins the frame body.
        while (true) {
            val line = readLine(ins) ?: return null
            if (line.isEmpty()) {
                if (contentLength > 0) break
                continue
            }
            val p = line.split(':', limit = 2)
            if (p.size == 2 && p[0].trim().equals("Content-Length", ignoreCase = true)) {
                contentLength = p[1].trim().toIntOrNull() ?: -1
                if (contentLength <= 0 || contentLength > MAX_FRAME) return null
            } else if (p.size == 2 &&
                p[0].trim().equals("X-Doorbell-Video-Rotation", ignoreCase = true)) {
                rotation = p[1].trim().toIntOrNull() ?: 0
            }
        }
        val buf = ByteArray(contentLength)
        var off = 0
        while (off < contentLength) {
            val n = ins.read(buf, off, contentLength - off)
            if (n < 0) return null
            off += n
        }
        return Frame(buf, ((rotation % 360) + 360) % 360)
    }

    /** Read one bounded CRLF-terminated ASCII line. */
    private fun readLine(ins: InputStream): String? {
        val sb = StringBuilder(64)
        while (true) {
            val c = ins.read()
            if (c < 0) return null
            if (c == '\n'.code) return sb.toString().trimEnd('\r')
            sb.append(c.toChar())
            if (sb.length > 512) return null
        }
    }

    companion object {
        private const val TAG = "doorbell-mjpeg"
        private const val MAX_FRAME = 4 * 1024 * 1024
        private const val MAX_WIDTH = 1280
        private const val MAX_HEIGHT = 720
    }
}
